libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1647,8 +1647,22 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		else if (functionName->name() == "suicide" && functionType->kind() == FunctionType::Kind::Selfdestruct)
 			m_errorReporter.typeError(_functionCall.location(), "\"suicide\" has been deprecated in favour of \"selfdestruct\"");
 	}
-	if (!m_insideEmitStatement && functionType->kind() == FunctionType::Kind::Event)
-		m_errorReporter.typeError(_functionCall.location(), "Event invocations have to be prefixed by \"emit\".");
+
+	// Declaring it here since it's used here and below again
+	bool const abiEncodeV2 = m_scope->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2);
+
+	if (functionType->kind() == FunctionType::Kind::Event)
+	{
+		if (!m_insideEmitStatement)
+			m_errorReporter.typeError(_functionCall.location(), "Event invocations have to be prefixed by \"emit\".");
+		if (!abiEncodeV2)
+			for (size_t i = 0; i < arguments.size(); ++i)
+			{
+				auto const& argType = type(*arguments[i]);
+				if (argType->category() == Type::Category::Struct)
+					m_errorReporter.typeError(arguments[i]->location(), "Structs cannot be used in events without the experimental ABIEncoderV2 feature.");
```

Comment:
Should add the helper message (we have some examples for this): "Use pragma experimental... to turn it on"

Reply:
Updated